New animations algorithm (Redux)

jmacato opened the pull request:
- What does the pull request do?
   - Revamps the current state machine to a much simpler, mathematical interpolation algorithm.
- What is the current behavior?
   - The current algorithm requires a complex state machine and is also frame-dependent, which can cause timing inaccuracies.

To-do list

- [x] Implement new animations algorithm.
- [x] Re-implement Play states.
- [x] Re-implement Transitions.

Commit:
Make ASM more Genericized.

Commit:
Remove underscores on ASM fields.

Commit:
Add new properties for Animations.

Commit:
Add the global frame count to AnimationStateTimer.

Commit:
Add the new animation algorithm.

Commit:
Optimize & remove excess numeric castings.

Commit:
Merge branch 'master' into new-animations-algo

Commit:
More optimizations.

Commit:
Merge branch 'new-animations-algo' of github.com:AvaloniaUI/Avalonia into new-animations-algo

Commit:
Fix onComplete assignment.

Commit:
Fix onComplete assignment.

Commit:
Merge branch 'new-animations-algo' of github.com:AvaloniaUI/Avalonia into new-animations-algo

Commit:
Fix bug on PageTransitions by adding 1 to the zeroth iteration frame count modulo.

Commit:
Make Transitions Bind strongly typed.

Commit:
Merge branch 'master' into new-animations-algo

grokys commented on src/Avalonia.Animation/Animation.cs:
```diff
@@ -75,6 +75,16 @@ private static Type GetAnimatorType(AvaloniaProperty property)
+        /// <summary>
+        /// Sets the behavior for having a delay between repeats for this animation.
+        /// </summary>
+        public bool DelayBetweenRepeats { get; set; }
```
A delay of how long? Might be worth explaining what this means a little in a `<remarks>` section of the xaml docs.

grokys commented on src/Avalonia.Animation/AnimatorStateMachine`1.cs:
```diff
@@ -1,273 +1,208 @@
+            if (playState == PlayState.Stop)
+                DoComplete();
 
             // Save state and pause the machine
```
I'd prefer not to have commented-out code in master as it quickly becomes stale, and it's not obvious _why_ it was commented out.

jmacato replied:
Sorry about that sir, still in progress and forgot to place a TODO. This will not be in the final PR

grokys commented on src/Avalonia.Animation/Timing.cs:
```diff
@@ -33,18 +36,23 @@ public static class Timing
 
             TransitionsTimer = globalTimer
-                               .Select(p => _transitionsFrameCount++)
+                               .Select(p => p)
```
I don't think is needed, right?

grokys reviewed (changes requested):
Looks good, and is a lot less code! Just a few small nits.

Am I right in thinking that the global play state hasn't been implemented in this PR? If so, it's worth mentioning that in the description.

grokys commented:
Assuming I'm right, and this doesn't implement a global play state - is there any chance we could completely remove all traces of the global play state? I don't think the way it's currently done is right, I think a system more like that used by WPF's [`AnimationClock`](https://msdn.microsoft.com/en-us/library/system.windows.media.animation.animationclock(v=vs.110).aspx) would be better where dependent animation clocks can created, and paused, independent of the global timer.

Commit:
Add PlayState support.
Redoing the main algorithm yet again.

Commit:
Merge branch 'new-animations-algo' of github.com:AvaloniaUI/Avalonia into new-animations-algo

Commit:
Make the algorithm use TimeSpans directly instead of converting to quantized framecounts to avoid/reduce quantization errors in interpolation.

Commit:
Fix Delay and Iteration Delay behaviors. AnimatorStateMachine is functionally complete.

Commit:
Merge branch 'master' into new-animations-algo

Commit:
Rename AnimatorStateMachine to AnimationsEngine to properly reflect the functions of the new algorithm.

Commit:
Implement a new timing engine for Transitions. Transtions are fixed now.Move out some stuff from Timing.cs.

jmacato commented:
@grokys Ready for another round of review, sir

Commit:
Merge branch 'master' into new-animations-algo

Commit:
Merge branch 'master' into new-animations-algo

Commit:
Merge branch 'master' into new-animations-algo

jkoritzinsky reviewed (approved):
LGTM!

jmacato commented:
Blockers (sorry forgot to mention this) 
1. There are some flickering on looping animations, mainly on the RenderDemo's Looping heart animation. Sir @grokys found out that the ScaleTransform is having a NaN value and im still not sure where's that coming from...
2. The TransitionsEngine attachment to the Timing code looks sound but there is still some flickering issues when any transitions are triggered...

jkoritzinsky committed:
Fix case where timing lands on exactly a new keyframe.

Commit:
Finally fixed the flickering heart bug!
Simplify the Keyframe selection and removed all LINQ code.
Use the Cue property on AnimatorKeyFrame as the time index and integrate the IsNeutral value as an internal property.

Commit:
Fix merge conflict

jmacato commented:
Blocker 1 is now resolved! After a month of running around cluelessly...

Commit:
Merge branch 'master' into new-animations-algo

grokys reviewed (changes requested):
This is looking great - a lot has been simplified! Just a few changes I think. You may also be interested in `SingleSubscriberObservableBase` as a base class for observables that can be subscribed just once - it handles some of the details (though it's not documented, sorry :( )

Commit:
Fix nits & polish up some code

Commit:
Merge branch 'master' into new-animations-algo

Commit:
Remove Epsilon equality checking since the animations
dont require such precision.

Commit:
Merge branch 'new-animations-algo' of github.com:AvaloniaUI/Avalonia into new-animations-algo

Commit:
Minor fix on TransitionsEngine

Commit:
Fix the transitions bug by pushing zero immediately after Subscription.
Thanks to sir @grokys for this fix.

Fix XML Doc on Animatable.cs.

Commit:
Fix naming on Animations & Transitions to better reflect their purpose.

Make RenderDemo & ControlCatalog's TabItems opacity transition faster.

jmacato commented:
Blocker 2 resolved! What a silly bug... 
Thanks to sir @grokys for the insight & fix. I think this is ready to be merged!!

grokys reviewed (approved):
LGTM!

Commit:
Make naming consistent.

Commit:
Make naming consistent.

Commit:
Merge branch 'new-animations-algo' of github.com:AvaloniaUI/Avalonia into new-animations-algo

Commit:
Merge branch 'master' into new-animations-algo
